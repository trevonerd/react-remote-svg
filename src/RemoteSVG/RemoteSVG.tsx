import React, { useState, useEffect } from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/react';
import { useInView } from 'react-intersection-observer';

import { imageVariants } from './utils/imageVariants';
import propsFilter from './utils/propFilter';

const supportsIntersectionObserver = typeof window !== 'undefined' && 'IntersectionObserver' in window;

export interface RemoteSVGProps extends React.HTMLAttributes<HTMLSpanElement> {
  activeEffect?: React.CSSProperties;
  alt?: string;
  dark?: boolean;
  disabledEffect?: React.CSSProperties;
  height?: number | string;
  hoverEffect?: React.CSSProperties;
  isActive?: boolean;
  isDisabled?: boolean;
  lazyLoad?: boolean;
  title?: string;
  url: string;
  useImageActive?: boolean;
  useImageDisabled?: boolean;
  useImageHover?: boolean;
  width?: number | string;
}

type StyledSVGProps = {
  $activeEffect?: React.CSSProperties;
  $disabledEffect?: React.CSSProperties;
  $hoverEffect?: React.CSSProperties;
  $isActive?: boolean;
  $isDisabled?: boolean;
  height: number | string;
  width: number | string;
};

const StyledSVG = React.memo(styled('img', propsFilter)<StyledSVGProps>`
  width: ${({ width }) => (typeof width === 'number' ? `${width}px` : width)};
  height: ${({ height }) => (typeof height === 'number' ? `${height}px` : height)};
  cursor: ${({ $hoverEffect, $isDisabled: isDisabled }) => ($hoverEffect && !isDisabled ? 'pointer' : 'inherit')};

  ${({ $isActive, $isDisabled, $activeEffect, $disabledEffect }) => {
    if ($isActive) {
      return { ...$activeEffect };
    }

    if ($isDisabled) {
      return { ...$disabledEffect };
    }
  }};

  ${({ $hoverEffect, $isDisabled: isDisabled }) =>
    $hoverEffect &&
    !isDisabled &&
    css`
      &:hover {
        ${{ ...$hoverEffect }}
      }
    `}
`);

const StyledSVGContainer = styled.span`
  align-items: center;
  display: flex;
  justify-content: center;
`;

const RemoteSVG: React.FC<RemoteSVGProps> = ({
  activeEffect,
  dark,
  disabledEffect,
  height = 24,
  hoverEffect,
  isActive = false,
  isDisabled = false,
  lazyLoad = supportsIntersectionObserver,
  title,
  alt = title,
  url,
  useImageActive,
  useImageDisabled,
  useImageHover,
  width = 24,
  ...rest
}) => {
  const [imgSrc, setImgSrc] = useState(url);
  const [isHovered, setIsHovered] = useState(false);
  const [variants, setVariants] = useState(imageVariants.light);
  const [isLoaded, setIsLoaded] = useState(false);

  const { ref, inView } = useInView({
    threshold: 0,
    rootMargin: '200px',
  });

  useEffect(() => {
    if (dark) {
      setVariants(imageVariants.dark);
    } else {
      setVariants(imageVariants.light);
    }
  }, [dark]);

  useEffect(() => {
    const loadImage = () => {
      setIsLoaded(true);
    };

    if (!lazyLoad || inView) {
      loadImage();
    }
  }, [lazyLoad, inView]);

  useEffect(() => {
    let newSrc = url;

    if (isHovered && useImageHover && !isDisabled) {
      newSrc = url.replace(/\.[^.]*$/, `${variants.hover}$&`);
    } else if (useImageDisabled) {
      newSrc = url.replace(/\.[^.]*$/, `${variants.disabled}$&`);
    } else if (useImageActive) {
      newSrc = url.replace(/\.[^.]*$/, `${variants.active}$&`);
    } else {
      // default variant
      newSrc = url.replace(/\.[^.]*$/, `${variants.normal}$&`);
    }

    setImgSrc(newSrc);
  }, [isHovered, url, dark, useImageHover, isDisabled, useImageDisabled, useImageActive, variants]);

  const handleMouseEnter = () => setIsHovered(true);
  const handleMouseLeave = () => setIsHovered(false);

  return (
    <StyledSVGContainer ref={ref}>
      {isLoaded ? (
        <StyledSVG
          $activeEffect={activeEffect}
          $disabledEffect={disabledEffect}
          $hoverEffect={hoverEffect}
          $isActive={isActive}
          $isDisabled={isDisabled}
          alt={alt}
          height={height}
          src={imgSrc}
          title={title}
          width={width}
          onMouseEnter={handleMouseEnter}
          onMouseLeave={handleMouseLeave}
          {...rest}
        />
      ) : (
        <span
          style={{
            width: typeof width === 'number' ? `${width}px` : width,
            height: typeof height === 'number' ? `${height}px` : height,
            display: 'inline-block',
          }}
        />
      )}
    </StyledSVGContainer>
  );
};

export default RemoteSVG;
