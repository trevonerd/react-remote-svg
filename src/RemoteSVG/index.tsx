export { default as RemoteSVG } from './RemoteSVG';
export type { RemoteSVGProps } from './RemoteSVG';

export { default as RemoteSVGMock } from './__mocks__/RemoteSVGMock';
