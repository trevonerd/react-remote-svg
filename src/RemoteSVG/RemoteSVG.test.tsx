import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
// eslint-disable-next-line @typescript-eslint/no-unused-vars, unused-imports/no-unused-imports
import RemoteSVG from './RemoteSVG';


const mockSVGContent = '<svg data-testid="remote-svg" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"></svg>';
const mockUrl = 'https://test-url.com/test.svg';

describe('RemoteSVG Component', () => {
  beforeEach(() => {
    (window.fetch as jest.Mock) = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        text: () => Promise.resolve(mockSVGContent),
      }),
    );

    (global.IntersectionObserver as jest.Mock) = jest
      .fn()
      .mockImplementation((callback: IntersectionObserverCallback) => {
        const entries = [
          {
            isIntersecting: true,
          },
        ];
        // @ts-expect-error IntersectionObserver is a mock
        setTimeout(() => callback(entries, null), 0);

        return {
          observe: jest.fn(),
          unobserve: jest.fn(),
          disconnect: jest.fn(),
        };
      });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('fetches svg and renders', async () => {
    const { asFragment } = render(<RemoteSVG url={mockUrl} />);

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders active SVG correctly', async () => {
    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        isActive={true}
        activeColor={'#f40000'}
        lazyLoad={true}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders disabled SVG correctly', async () => {
    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        isDisabled={true}
        disabledColor={'#333333'}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders SVG with hover correctly', async () => {
    const mockUrl = 'https://test-url.com/test.svg';

    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        hoverColor={'#f67645'}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('fetches SVG when not available in cache', async () => {
    const mockUrl = 'https://test-url.com/test.svg';

    (global.fetch as jest.Mock) = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        text: () => Promise.resolve(mockSVGContent),
      }),
    );

    const { asFragment } = render(<RemoteSVG url={mockUrl} />);

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders SVG correctly with custom width and height as strings', async () => {
    const mockUrl = 'https://test-url.com/test.svg';

    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        width={'100px'}
        height={'100px'}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders SVG correctly with custom width and height as numbers', async () => {
    const mockUrl = 'https://test-url.com/test.svg';

    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        width={100}
        height={100}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders active SVG correctly without active color', async () => {
    const mockUrl = 'https://test-url.com/test.svg';

    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        isActive={true}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders disabled SVG correctly without active color', async () => {
    const mockUrl = 'https://test-url.com/test.svg';

    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        isDisabled={true}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('no url provided', async () => {
    const mockUrl = '';

    const { asFragment } = render(
      <RemoteSVG
        url={mockUrl}
        isDisabled={true}
      />,
    );

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('fetches svg and renders if intersectionObserver is undefined', async () => {
    const originalIntersectionObserver = window.IntersectionObserver;
    // @ts-expect-error deleting window.IntersectionObserver
    delete window.IntersectionObserver;

    const mockUrl = 'https://test-url.com/test.svg';

    const { asFragment } = render(<RemoteSVG url={mockUrl} />);

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);

      const svgElement = screen.getByTestId('remote-svg');
      expect(svgElement).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();

    window.IntersectionObserver = originalIntersectionObserver;
  });

  it('handles fetch error correctly', async () => {
    const mockUrl = 'https://test-url.com/test.svg';

    (window.fetch as jest.Mock) = jest.fn().mockImplementation(() => Promise.reject('Network error'));

    global.console = { error: jest.fn() } as unknown as Console;

    const { asFragment } = render(<RemoteSVG url={mockUrl} />);

    await waitFor(() => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith(mockUrl);
      expect(console.error).toBeCalledWith('Error fetching SVG:', mockUrl, 'Network error');
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
