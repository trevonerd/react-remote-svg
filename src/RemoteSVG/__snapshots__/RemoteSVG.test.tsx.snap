// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteSVG Component dark mode - loads the correct image variant if isActive 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_da.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component dark mode - loads the correct image variant if isActive with activeEffect 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
  -webkit-filter: brightness(0.8);
  filter: brightness(0.8);
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_da.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component dark mode - loads the correct image variant if isDisabled 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_dd.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component dark mode - loads the correct image variant if isDisabled with disabledEffect 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
  -webkit-filter: contrast(0.8);
  filter: contrast(0.8);
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_dd.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component loads the correct image variant if isActive 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_a.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component loads the correct image variant if isActive with activeEffect 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
  -webkit-filter: brightness(0.8);
  filter: brightness(0.8);
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_a.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component loads the correct image variant if isDisabled 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_d.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component loads the correct image variant if isDisabled with disabledEffect 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
  -webkit-filter: contrast(0.8);
  filter: contrast(0.8);
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="image_d.png"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component renders SVG 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
}

<span
    class="emotion-0"
  >
    <img
      alt="test image"
      class="emotion-1"
      height="24"
      src="https://test-url.com/test.svg"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component renders SVG with correct alt and title attributes 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
}

<span
    class="emotion-0"
  >
    <img
      alt="test image alt"
      class="emotion-1"
      height="24"
      src="https://test-url.com/test.svg"
      title="test image title"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component renders SVG with hover correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: pointer;
}

.emotion-1:hover {
  -webkit-filter: brightness(0.8);
  filter: brightness(0.8);
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="https://test-url.com/test.svg"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component renders active SVG correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

<span
    class="emotion-0"
  >
    <span
      style="width: 24px; height: 24px; display: inline-block;"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component renders disabled SVG correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  width: 24px;
  height: 24px;
  cursor: inherit;
  opacity: 0.5;
}

<span
    class="emotion-0"
  >
    <img
      class="emotion-1"
      height="24"
      src="https://test-url.com/test.svg"
      width="24"
    />
  </span>
</DocumentFragment>
`;

exports[`RemoteSVG Component should not renders SVG with active, disabled effect if isActive or isDisabled props are false 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

<span
    class="emotion-0"
  >
    <span
      style="width: 24px; height: 24px; display: inline-block;"
    />
  </span>
</DocumentFragment>
`;
